http-gateways/IPLD_GATEWAY.md

```diff
@@ -0,0 +1,196 @@
+- `application/vnd.ipld.dag-json`: Return the block specified by the path encoded in `dag-json`.
+- `application/vnd.ipld.dag-cbor`: Return the block specified by the path ecnoded in `dag-cbor`.
+
+If no `Accept` header is present in the request, it will be assumed to be `application/vnd.ipld.dag-json`.
+
+#### `Content-Type` (request header)
+
+This header applies to `PUT/POST/PATCH` requests for `/ipld/*` paths on IPLD gateways which also support the [writable gateways spec](./WRITABLE_GATEWAY.md).
+Including it will hint to the writable IPLD gateway which encoding to use to parse the request body into the IPLD Data Model.
+
+- `application/json`: Interpret in the same way as `application/vnd.ipld.dag-json`.
+- `application/vnd.ipld.dag-json`: Return the block specified by the path encoded in `dag-json`.
+- `application/vnd.ipld.dag-cbor`: Return the block specified by the path ecnoded in `dag-cbor`.
+
+### Request Query Parameters
+
+#### `format` (request query parameter)
+
+Optional, `format=<format>` can be used to request specific encodings.
+
+This is a URL-friendly alternative to sending `Accept: application/vnd.ipld.<format>` header, see [Accept](#accept-request-header) for more details.
+
+### Path Segments and Path Segment Parameters
+
+Path segments are used for IPLD data model [traversal](https://ipld.io/docs/data-model/traversal/).
+Each segment is separated by a `/` and contains a utf8 `name` followed by an optional set of parameters using the [Matrix URI format](https://www.w3.org/DesignIssues/MatrixURIs.html).
+
+Example:
+
+```
+/ipld/bafywhatever/foo/bar;extra=thing;whatever=here/
+```
+
+Path segments can have additional parameters added to them by separating them using semicolons (`;`) and having key-value pairs separated by an `=`.
+This format is based on the [Matrix URI proposal from the w3c](https://www.w3.org/DesignIssues/MatrixURIs.html).
+Note that these parameters are stripped from the segment name when passed to any underlying traversal code.
+
+This spec only perscribes two reserved parameter names: `ADL` for specifying [Advanced Data Layouts](https://ipld.io/docs/advanced-data-layouts/) to process the data with, and `schema` to specify an [IPLD schema](https://ipld.io/docs/schemas/intro/) to use to interpret the data.
+Other names may be specified in future specs that build upon this one.
+
+#### ADL (segment parameter)
+
+Segments may contain an `adl` key which points to a name of an [Advanced Data Layout](https://ipld.io/docs/advanced-data-layouts/intro/) to process the node with.
+
+The supported ADL names will vary based on gateway.
+
+An example value would be `;adl=hamt` to specify the [HAMT](https://ipld.io/specs/advanced-data-layouts/hamt/) ADL that's used to represent large maps.
+
+For example `/ipld/bafyreic672jz6huur4c2yekd3uycswe2xfqhjlmtmm5dorb6yoytgflova;adl=hamt/yes` (taken from the HAMT examples), should resolve to the following:
```

Comment:
Why `hamt/yes` and not `hamt` ?

Reply:
The `yes` is another path segment in the URL. This shows how the ADL can be used at an arbitrary point in the traversal and not just at the start or end.

Should I add wording to explain that?